File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys
input = sys.stdin.readline

n = int(input())
triangle = [list(map(int, input().split())) for _ in range(n)]

dp = [[0] * i for i in range(1, n + 1)]

dp[0][0] = triangle[0][0]

for i in range(1, n):
    for j in range(i + 1):
        if j == 0:
            dp[i][j] = dp[i-1][j] + triangle[i][j]
        elif j == len(dp[i]) - 1:
            dp[i][j] = dp[i - 1][-1] + triangle[i][j]
        else:
            left = dp[i-1][j-1] + triangle[i][j]
            right = dp[i-1][j] + triangle[i][j]
            dp[i][j] = max(left, right)
            
print(max(dp[n-1]))

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    // dp[i][j][0, 0] 왼쪽에서 내려온 경우, 오른쪽에서 내려온 경우
    const dp = Array.from({length:triangle.length}, (_,i) => Array.from({length:i+1}, (_, j) => 0));
    dp[0][0] = triangle[0][0];
    
    for (let i = 1; i < triangle.length; i++) {
        for (let j = 0; j < i+1; j++) {
            if (j === 0) {
                dp[i][j] = dp[i-1][0] + triangle[i][j];
            }
            else if (j === i) {
                dp[i][j] = dp[i-1][i-1] + triangle[i][j];
            }
            else {
                dp[i][j] = triangle[i][j] + Math.max(dp[i-1][j-1], dp[i-1][j]);
            }
        }
    }
    return Math.max(...dp[triangle.length - 1])
}